### PR TITLE
PCIe Monitor service

### PR DESCRIPTION
**Pull request by sujinmkang**
PCIe Monitor Service which will monitor the PCIe device (and bus) status periodically.

### CODE REVIEW

File: doc/pcie-mon/pcie-daemon-hld.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+## About This Manual ##
+
+This document is intend to give the idea of how to monitor the platform PCIe devices and alert any problem on PCIe buses and devices on SONiC using pcie-mon service and pcied on PMON container.
+
+
+## 1. PCIe Monitor service design ##
+
+New PCIe Monitor service is designed to use the PcieUtil utility to check the current status of PCIe devices and buses and alert if there is any missing devices or any error while communicating on the PCIe buses.
+
+PCIe device monitoring will be done in two separate services, `pcie-mon.service` which is a systemd service, will monitor the PCIe device during the boot time and `pcied` which is a daemon in PMON container will monitor during the runtime.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-daemon-hld.md:21]: Now that the systemd service is a one-time check and not a periodic monitor (as the monitoring is now done by the PMon daemon), I suggest renaming it from "pcie-mon.service" to something else, like 'pcie-check.service".

File: doc/pcie-mon/pcie-daemon-hld.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+## About This Manual ##
+
+This document is intend to give the idea of how to monitor the platform PCIe devices and alert any problem on PCIe buses and devices on SONiC using pcie-chk service and pcied on PMON container.
+
+
+## 1. PCIe Monitor service design ##
+
+New PCIe Monitor service is designed to use the PcieUtil utility to check the current status of PCIe devices and buses and alert if there is any missing devices or any error while communicating on the PCIe buses.
+
+PCIe device monitoring will be done in two separate services, `pcie-chk.service` which is a systemd service, will monitor the PCIe device during the boot time and `pcied` which is a daemon in PMON container will monitor during the runtime.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-daemon-hld.md:21]: Suggest using the full word "check", as it's only two more chars (i.e., "pcie-check.service")

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+### Revision
+ | Rev |     Date    |       Author       | Change Description                             |
+ |:---:|:-----------:|:------------------:|------------------------------------------------|
+ | 0.1 |             |     Sujin Kang     | Initial version                                |
+ | 0.2 |             |     Sujin Kang     | Add rescan for pcie device missing during boot | 
+ |     |             |                    | Add pcied to PMON for runtime monitoring       |
+
+## About This Manual ##
+
+This document is intend to give the idea of how to monitor the platform PCIe devices and alert any problem on PCIe buses and devices on SONiC using pcie-check service and pcied on PMON container.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:14]: s/intend/intended/
s/alert any problem/alert of any problems/

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+
+## 1. PCIe Monitor service design ##
+
+New PCIe Monitor service is designed to use the PcieUtil utility to check the current status of PCIe devices and buses and alert if there is any missing devices or any error while communicating on the PCIe buses.
+
+PCIe device monitoring will be done in two separate services, `pcie-check.service` which is a systemd service, will monitor the PCIe device during the boot time and `pcied` which is a daemon in PMON container will monitor during the runtime.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:21]: >`pcie-check.service` which is a systemd service, will monitor the PCIe device during the boot time

This service only performs one check, it doesn't really monitor, so I suggest replacing "monitor" with "check"

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+First, pcie-check.service will be added to check the pcie device enumeration status, trigger the pci device rescan if there is any missing device and indicate any device missing to the party that are interested in the device enumeration, for example, kernel_bde driver, platform drivers and etc.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:23]: Should also mention that pcie-check.service will attempt to trigger a PCIe bus rescan if any required device(s) are not present.

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        if(status):
+            for key, value in device_dict.iteritems():
+                print("Device on PCIe bus: %s" was %s" % (key, value))
+                 
+### 1.4 PCIe Check Service `pcie-check.service` flow ###
+
+pcie-check.service will be started by systemd during boot up and it will spawn a thread to check PCIe device status and perform the rescan pci devices if there is any missing devices after rc.local.service is completed and it will update the state db with pcie device satus after the `pcieutil pcie-chek` call so that the dependent services/container or kernel driver can be started or stopped based on the status.
+
+Detailed flow as showed in below chart: 
+![](https://github.com/Azure/SONiC/blob/master/images/pcie-check.svg)
```
Comment (padmanarayana) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:95]: Unable to view this file.

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+
+### 1.5 PCIe daemon `pcied` flow ###
+
+pcied will be started by PMON container will continue monitoring the PCIe device status during run time and it will check the PCIe device status periodically every 1 minute and update the state db when the status is checked.
+
+Detailed flow as showed in below chart:
+![](https://github.com/Azure/SONiC/blob/master/images/pcied.svg)
```
Comment (padmanarayana) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:103]: unable to view this file..

File: doc/pcie-mon/pcie-monitoring-services-hld.md (hunk 6 above)
Comment (paavaanan) [reply doc/pcie-mon/pcie-monitoring-services-hld.md:95]: https://github.com/Azure/SONiC/blob/70a152f1b98e145c9f0771e7cda7a951d98a978e/images/pcie-check.svg --- This link needs to be updated

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+This document is intended to give the idea of how to monitor the platform PCIe devices and alert of any problems on PCIe buses and devices on SONiC using pcie-check service and pcied on PMON container.
+
+
+## 1. PCIe Monitor service design ##
+
+New PCIe Monitor service is designed to use the PcieUtil utility to check the current status of PCIe devices and buses and alert if there is any missing devices or any error while communicating on the PCIe buses.
+
+PCIe device monitoring will be done in two separate services, `pcie-check.service` which is a systemd service, will check the PCIe device during the boot time and `pcied` which is a daemon in PMON container will monitor during the runtime.
+
+First, pcie-check.service will be added to check the pcie device enumeration status, trigger 15 maximum retries of a pci device rescan if there is any missing device and save the result status of pcie device check into the STATE_DB to indicate any device missing to the party that are interested in the device enumeration, for example, kernel_bde driver, platform drivers and etc.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:23]: You are no longer triggering a maximum of 15 retries, correct?

Comment (sujinmkang) [reply doc/pcie-mon/pcie-monitoring-services-hld.md:23]: good catch!

File: doc/pcie-mon/pcie-monitoring-services-hld.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+2. `PcieUtil` will provide APIs `load_config_file`, `get_pcie_device` and `get_pcie_check` to get the expected PCIe device list and informations, to get the current PCIe device information, and check if any PCIe device is missing or if there is any PCIe bus error.
+
+![pcieinfo_design](https://github.com/Azure/SONiC/blob/master/doc/pcieinfo_design.md)
+
+### 1.2 PCIe device configuration file ###
+
+PcieUtil needs to get the expected PCIe device information to check the PCIe device status periodically, which is different for each platform/hardware sku.
+
+Each vendor need to generate the PCIe device configuration file name as pcie.yml and locate the file under device/<platform>/<hardware_skus>/plugins.
```
Comment (jleveque) [inline doc/pcie-mon/pcie-monitoring-services-hld.md:45]: The angle brackets need to be escaped to render properly in markdown:

```
device//<platform/>//<hardware_skus/>/plugins
```

**Pull request merged**